client/asset/dcr/dcr.go

```diff
@@ -2925,50 +2859,19 @@ func (dcr *ExchangeWallet) getBestBlock(ctx context.Context) (*block, error) {
 func (dcr *ExchangeWallet) mainchainAncestor(ctx context.Context, blockHash *chainhash.Hash) (*chainhash.Hash, int64, error) {
 	checkHash := blockHash
 	for {
-		checkBlock, err := dcr.wallet.GetBlockHeaderVerbose(ctx, checkHash)
+		checkBlock, isMainchain, err := dcr.getMainchainHeader(checkHash)
 		if err != nil {
 			return nil, 0, fmt.Errorf("getblockheader error for block %s: %w", checkHash, err)
 		}
-		if checkBlock.Confirmations > -1 {
+		if isMainchain {
 			// This is a mainchain block, return the hash and height.
 			return checkHash, int64(checkBlock.Height), nil
 		}
 		if checkBlock.Height == 0 {
-			return nil, 0, fmt.Errorf("no mainchain ancestor for block %s", blockHash.String())
+			return nil, 0, fmt.Errorf("no mainchain ancestor for block %s", blockHash)
 		}
-		checkHash, err = chainhash.NewHashFromStr(checkBlock.PreviousHash)
-		if err != nil {
-			return nil, 0, fmt.Errorf("error decoding previous hash %s for block %s: %w",
-				checkBlock.PreviousHash, checkHash.String(), err)
-		}
-	}
-}
-
-func (dcr *ExchangeWallet) isMainchainBlock(ctx context.Context, block *block) (bool, error) {
-	blockHeader, err := dcr.wallet.GetBlockHeaderVerbose(ctx, block.hash)
-	if err != nil {
-		return false, fmt.Errorf("getblockheader error for block %s: %w", block.hash, err)
-	}
-	// First validation check.
-	if blockHeader.Confirmations < 0 || int64(blockHeader.Height) != block.height {
-		return false, nil
-	}
-	// Check if the next block invalidated this block's regular tree txs.
-	// This block checks out if there is no following block yet.
-	if blockHeader.NextHash == "" {
-		return true, nil
-	}
-	nextBlockHash, err := chainhash.NewHashFromStr(blockHeader.NextHash)
-	if err != nil {
-		return false, fmt.Errorf("block %s has invalid nexthash value %s: %v",
-			block.hash, blockHeader.NextHash, err)
-	}
-	nextBlockHeader, err := dcr.wallet.GetBlockHeaderVerbose(ctx, nextBlockHash)
-	if err != nil {
-		return false, fmt.Errorf("getblockheader error for block %s: %w", nextBlockHash, err)
+		checkHash = &checkBlock.PrevBlock
 	}
-	validated := nextBlockHeader.VoteBits&1 != 0
```

Comment:
To reduce the possibility of alternative `Wallet` implementations bombing themselves, we should include in the `IsMainchain` doc the need to check the next header vote bits.